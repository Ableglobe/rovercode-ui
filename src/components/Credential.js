import React, { Component } from 'react';
import {
  Button,
  Form,
  Icon,
  Modal,
} from 'semantic-ui-react';
import { hot } from 'react-hot-loader';
import PropTypes from 'prop-types';
import fileDownload from 'js-file-download';

const defaultState = {
  downloadOpen: false,
  apName: null,
  apPassword: null,
};

class Credential extends Component {
  constructor(props) {
    super(props);

    this.state = defaultState;
  }

  downloadButton = () => (
    <Button onClick={this.handleDownloadOpen}>
      <Icon name="download" />
      Download Credentials
    </Button>
  )

  handleDownloadOpen = () => this.setState({ downloadOpen: true })

  handleDownloadClose = () => this.setState({ downloadOpen: false })

  handleNameChange = e => this.setState({ apName: e.target.value })

  handlePasswordChange = e => this.setState({ apPassword: e.target.value })

  downloadCredentials = () => {
    const { rover } = this.props;
    const { apName, apPassword } = this.state;

    const credentials = {
      CLIENT_ID: rover.client_id,
      CLIENT_SECRET: rover.client_secret,
      AP_NAME: apName,
      AP_PASSWORD: apPassword,
    };
    let fileContents = '';
    Object.keys(credentials).forEach((key) => {
      fileContents += `${key}=${credentials[key]}\n`;
    });
    fileDownload(fileContents, `rovercode_${rover.name}.env`);

    this.setState(defaultState);
  }

  render() {
    const { downloadOpen } = this.state;

    return (
      <Modal trigger={this.downloadButton()} open={downloadOpen} onClose={this.handleDownloadClose}>
        <Modal.Header>
          Enter WiFi Credentials
        </Modal.Header>
        <Modal.Content>
          <p>
            The credentials entered are only used for the configuration file for the rover.
            This information does not leave the local browser and is not sent to the server.
          </p>
          <Form id="apForm" onSubmit={this.downloadCredentials}>
            <Form.Input placeholder="Access point name" onChange={this.handleNameChange} required />
            <Form.Input type="password" placeholder="Access point password" onChange={this.handlePasswordChange} required />
          </Form>
        </Modal.Content>
        <Modal.Actions>
          <Button primary type="submit" form="apForm">
            Download
          </Button>
          <Button onClick={this.handleDownloadClose}>
            Cancel
          </Button>
        </Modal.Actions>
      </Modal>
    );
  }
}

Credential.propTypes = {
  rover: PropTypes.shape({
    name: PropTypes.string.isRequired,
    client_id: PropTypes.string.isRequired,
    client_secret: PropTypes.string.isRequired,
  }).isRequired,
};

export default hot(module)(Credential);
