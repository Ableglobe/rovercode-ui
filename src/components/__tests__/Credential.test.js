import React from 'react';
import { Button, Form } from 'semantic-ui-react';
import { shallowWithIntl } from 'enzyme-react-intl';

import fileDownload from 'js-file-download';

import Credential from '../Credential';

jest.mock('js-file-download');

const rover = {
  id: 1,
  name: 'Sparky',
  config: {
    left_eye_port: 1,
    right_eye_port: 2,
  },
  client_id: '1234',
  client_secret: '5678',
};

describe('The Credential component', () => {
  beforeEach(() => {
    fileDownload.mockClear();
  });

  test('renders on the page with no errors', () => {
    const wrapper = shallowWithIntl(<Credential rover={rover} />).dive();
    expect(wrapper).toMatchSnapshot();
  });

  test('operates dialog properly', () => {
    const wrapper = shallowWithIntl(<Credential rover={rover} />).dive();

    wrapper.instance().handleDownloadOpen();

    expect(wrapper.state('downloadOpen')).toBe(true);

    wrapper.find(Form.Input).first().simulate('change', {
      target: {
        value: 'AP Name',
      },
    });
    wrapper.find(Form.Input).last().simulate('change', {
      target: {
        value: 'AP Password',
      },
    });

    expect(wrapper.state('apName')).toBe('AP Name');
    expect(wrapper.state('apPassword')).toBe('AP Password');

    wrapper.find(Form).simulate('submit');

    expect(wrapper.state('downloadOpen')).toBe(false);
    expect(fileDownload).toHaveBeenCalledWith(
      `CLIENT_ID=${rover.client_id}\nCLIENT_SECRET=${rover.client_secret}\nAP_NAME=AP Name\nAP_PASSWORD=AP Password\n`,
      `rovercode_${rover.name}.env`,
    );
  });

  test('cancel dialog does not download credentials', () => {
    const wrapper = shallowWithIntl(<Credential rover={rover} />).dive();

    wrapper.instance().handleDownloadOpen();

    expect(wrapper.state('downloadOpen')).toBe(true);

    wrapper.find(Button).last().simulate('click');

    expect(wrapper.state('downloadOpen')).toBe(false);
    expect(fileDownload).not.toHaveBeenCalled();
  });
});
