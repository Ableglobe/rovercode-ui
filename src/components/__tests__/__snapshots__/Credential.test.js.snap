// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Credential component renders on the page with no errors 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Credential
    rover={
      Object {
        "client_id": "1234",
        "client_secret": "5678",
        "config": Object {
          "left_eye_port": 1,
          "right_eye_port": 2,
        },
        "id": 1,
        "name": "Sparky",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "centered": true,
      "children": Array [
        <ModalHeader>
          Enter WiFi Credentials
        </ModalHeader>,
        <ModalContent>
          <p>
            The credentials entered are only used for the configuration file for the rover. This information does not leave the local browser and is not sent to the server.
          </p>
          <Form
            as="form"
            id="apForm"
            onSubmit={[Function]}
          >
            <FormInput
              as={[Function]}
              control={[Function]}
              onChange={[Function]}
              placeholder="Access point name"
              required={true}
            />
            <FormInput
              as={[Function]}
              control={[Function]}
              onChange={[Function]}
              placeholder="Access point password"
              required={true}
              type="password"
            />
          </Form>
        </ModalContent>,
        <ModalActions>
          <Button
            as="button"
            form="apForm"
            primary={true}
            type="submit"
          >
            Download
          </Button>
          <Button
            as="button"
            onClick={[Function]}
          >
            Cancel
          </Button>
        </ModalActions>,
      ],
      "closeOnDimmerClick": true,
      "closeOnDocumentClick": false,
      "dimmer": true,
      "eventPool": "Modal",
      "onClose": [Function],
      "open": false,
      "trigger": <Button
        as="button"
        onClick={[Function]}
      >
        <Icon
          as="i"
          name="download"
        />
        Download Credentials
      </Button>,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Enter WiFi Credentials",
        },
        "ref": null,
        "rendered": "Enter WiFi Credentials",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <p>
              The credentials entered are only used for the configuration file for the rover. This information does not leave the local browser and is not sent to the server.
            </p>,
            <Form
              as="form"
              id="apForm"
              onSubmit={[Function]}
            >
              <FormInput
                as={[Function]}
                control={[Function]}
                onChange={[Function]}
                placeholder="Access point name"
                required={true}
              />
              <FormInput
                as={[Function]}
                control={[Function]}
                onChange={[Function]}
                placeholder="Access point password"
                required={true}
                type="password"
              />
            </Form>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "The credentials entered are only used for the configuration file for the rover. This information does not leave the local browser and is not sent to the server.",
            },
            "ref": null,
            "rendered": "The credentials entered are only used for the configuration file for the rover. This information does not leave the local browser and is not sent to the server.",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "form",
              "children": Array [
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  onChange={[Function]}
                  placeholder="Access point name"
                  required={true}
                />,
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  onChange={[Function]}
                  placeholder="Access point password"
                  required={true}
                  type="password"
                />,
              ],
              "id": "apForm",
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "as": [Function],
                  "control": [Function],
                  "onChange": [Function],
                  "placeholder": "Access point name",
                  "required": true,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "as": [Function],
                  "control": [Function],
                  "onChange": [Function],
                  "placeholder": "Access point password",
                  "required": true,
                  "type": "password",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Button
              as="button"
              form="apForm"
              primary={true}
              type="submit"
            >
              Download
            </Button>,
            <Button
              as="button"
              onClick={[Function]}
            >
              Cancel
            </Button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "button",
              "children": "Download",
              "form": "apForm",
              "primary": true,
              "type": "submit",
            },
            "ref": null,
            "rendered": "Download",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "button",
              "children": "Cancel",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Cancel",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "centered": true,
        "children": Array [
          <ModalHeader>
            Enter WiFi Credentials
          </ModalHeader>,
          <ModalContent>
            <p>
              The credentials entered are only used for the configuration file for the rover. This information does not leave the local browser and is not sent to the server.
            </p>
            <Form
              as="form"
              id="apForm"
              onSubmit={[Function]}
            >
              <FormInput
                as={[Function]}
                control={[Function]}
                onChange={[Function]}
                placeholder="Access point name"
                required={true}
              />
              <FormInput
                as={[Function]}
                control={[Function]}
                onChange={[Function]}
                placeholder="Access point password"
                required={true}
                type="password"
              />
            </Form>
          </ModalContent>,
          <ModalActions>
            <Button
              as="button"
              form="apForm"
              primary={true}
              type="submit"
            >
              Download
            </Button>
            <Button
              as="button"
              onClick={[Function]}
            >
              Cancel
            </Button>
          </ModalActions>,
        ],
        "closeOnDimmerClick": true,
        "closeOnDocumentClick": false,
        "dimmer": true,
        "eventPool": "Modal",
        "onClose": [Function],
        "open": false,
        "trigger": <Button
          as="button"
          onClick={[Function]}
        >
          <Icon
            as="i"
            name="download"
          />
          Download Credentials
        </Button>,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Enter WiFi Credentials",
          },
          "ref": null,
          "rendered": "Enter WiFi Credentials",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <p>
                The credentials entered are only used for the configuration file for the rover. This information does not leave the local browser and is not sent to the server.
              </p>,
              <Form
                as="form"
                id="apForm"
                onSubmit={[Function]}
              >
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  onChange={[Function]}
                  placeholder="Access point name"
                  required={true}
                />
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  onChange={[Function]}
                  placeholder="Access point password"
                  required={true}
                  type="password"
                />
              </Form>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "The credentials entered are only used for the configuration file for the rover. This information does not leave the local browser and is not sent to the server.",
              },
              "ref": null,
              "rendered": "The credentials entered are only used for the configuration file for the rover. This information does not leave the local browser and is not sent to the server.",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "form",
                "children": Array [
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    onChange={[Function]}
                    placeholder="Access point name"
                    required={true}
                  />,
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    onChange={[Function]}
                    placeholder="Access point password"
                    required={true}
                    type="password"
                  />,
                ],
                "id": "apForm",
                "onSubmit": [Function],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "as": [Function],
                    "control": [Function],
                    "onChange": [Function],
                    "placeholder": "Access point name",
                    "required": true,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "as": [Function],
                    "control": [Function],
                    "onChange": [Function],
                    "placeholder": "Access point password",
                    "required": true,
                    "type": "password",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Button
                as="button"
                form="apForm"
                primary={true}
                type="submit"
              >
                Download
              </Button>,
              <Button
                as="button"
                onClick={[Function]}
              >
                Cancel
              </Button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "children": "Download",
                "form": "apForm",
                "primary": true,
                "type": "submit",
              },
              "ref": null,
              "rendered": "Download",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "children": "Cancel",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Cancel",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
